store/tikv/txn_scheduler.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv/latch"
+	binlog "github.com/pingcap/tipb/go-binlog"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type txnCommitter struct {
+	twoPC *twoPhaseCommitter
+	ch    chan error
+	ctx   context.Context
+	lock  *latch.Lock
+}
+
+func newTxnCommiter(ctx context.Context, committer *twoPhaseCommitter) *txnCommitter {
+	return &txnCommitter{
+		twoPC: committer,
+		ctx:   ctx,
+	}
+}
+
+func (tc *txnCommitter) commit(stale bool) uint64 {
+	if stale {
+		err := errors.Errorf("startTS %d is stale", tc.twoPC.startTS)
+		log.Debug(tc.twoPC.connID, err)
+		tc.ch <- errors.Annotate(err, txnRetryableMark)
+		return uint64(0)
+	}
+
+	commitTS, err := tc.execute()
+	tc.ch <- errors.Trace(err)
+	log.Debug(tc.twoPC.connID, " finish txn with startTS:", tc.twoPC.startTS, " commitTS:", commitTS, " error:", err)
+	return commitTS
+}
+
+func (tc *txnCommitter) execute() (commitTS uint64, err error) {
+	err = tc.twoPC.execute(tc.ctx)
+	if err != nil {
+		tc.twoPC.writeFinishBinlog(binlog.BinlogType_Rollback, 0)
+	} else {
+		commitTS = tc.twoPC.commitTS
+		tc.twoPC.writeFinishBinlog(binlog.BinlogType_Commit, int64(commitTS))
+	}
+	return
+}
+
+type txnMap struct {
+	txns map[uint64]*txnCommitter
+	sync.RWMutex
+}
+
+func (t *txnMap) put(startTS uint64, txn *txnCommitter) {
+	t.Lock()
+	defer t.Unlock()
+	if t.txns == nil {
+		t.txns = make(map[uint64]*txnCommitter)
+	}
+	if _, ok := t.txns[startTS]; ok {
+		panic(fmt.Sprintf("The startTS %d shouldn't be used in two transaction", txn.twoPC.startTS))
+	}
+	t.txns[startTS] = txn
+}
+
+func (t *txnMap) get(startTS uint64) *txnCommitter {
+	t.RLock()
+	defer t.RUnlock()
+	return t.txns[startTS]
+}
+
+func (t *txnMap) delete(startTS uint64) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.txns, startTS)
+}
+
+// txnScheduler is the scheduler to commit transactions.
+type txnScheduler struct {
+	latches *latch.Latches
+	txnMaps []txnMap
```

Comment:
Do we really need an array of `txnMap`? Maybe try try `sync.Map`?

From the document of `sync.Map`:
> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

It looks like our scenario fits the description of case 2.